packages/block-editor/src/hooks/text-align.js

```diff
@@ -0,0 +1,214 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { addFilter } from '@wordpress/hooks';
+import { getBlockSupport, hasBlockSupport } from '@wordpress/blocks';
+import { alignLeft, alignRight, alignCenter } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import { AlignmentControl, BlockControls } from '../components';
+import { useBlockEditingMode } from '../components/block-editing-mode';
+import { cleanEmptyObject, shouldSkipSerialization } from './utils';
+import { TYPOGRAPHY_SUPPORT_KEY } from './typography';
+
+const TEXT_ALIGN_SUPPORT_KEY = 'typography.textAlign';
+
+const TEXT_ALIGNMENT_OPTIONS = [
+	{
+		icon: alignLeft,
+		title: __( 'Align text left' ),
+		align: 'left',
+	},
+	{
+		icon: alignCenter,
+		title: __( 'Align text center' ),
+		align: 'center',
+	},
+	{
+		icon: alignRight,
+		title: __( 'Align text right' ),
+		align: 'right',
+	},
+];
+
+const VALID_TEXT_ALIGNMENTS = [ 'left', 'center', 'right' ];
+
+/**
+ * Returns the valid text alignments.
+ * Takes into consideration the text aligns supported by a block.
+ * Exported just for testing purposes, not exported outside the module.
+ *
+ * @param {?boolean|string[]} blockTextAlign Text aligns supported by the block.
+ *
+ * @return {string[]} Valid text alignments.
+ */
+export function getValidTextAlignments( blockTextAlign ) {
+	let validTextAlignments;
+	if ( Array.isArray( blockTextAlign ) ) {
+		validTextAlignments = VALID_TEXT_ALIGNMENTS.filter( ( textAlign ) =>
+			blockTextAlign.includes( textAlign )
+		);
+	} else if ( blockTextAlign === true ) {
+		// `true` includes all alignments...
+		validTextAlignments = [ ...VALID_TEXT_ALIGNMENTS ];
+	} else {
+		validTextAlignments = [];
+	}
+
+	return validTextAlignments;
+}
+
+/**
+ * Filters registered block settings, extending attributes to include `textAlign`.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Filtered block settings.
+ */
+export function addAttribute( settings ) {
+	// Allow blocks to specify their default style if needed.
+	if ( settings.attributes?.style?.default?.typography?.textAlign ) {
+		return settings;
+	}
+	if ( hasBlockSupport( settings, TEXT_ALIGN_SUPPORT_KEY ) ) {
+		// Gracefully handle if settings.attributes is undefined.
+		settings.attributes = {
+			...settings.attributes,
+			style: {
+				...settings.attributes?.style,
+				typography: {
+					...settings.attributes?.style?.typography,
+					textAlign: undefined,
```

Comment:
This is the only part that I can't explain well. When supporting _only_ textAlign, like the custom block shown in the Testing Instructions, the text-align setting was not saved as an attribute unless you explicitly defined a style object with this property.

Reply:
That does seem a bit odd 🤔 

I'll give this another review this week but in the meantime, I think several other places might need to be updated to be aware of this new support. 

Some might only need updating if we expose the text alignment control in Global Styles for block types which I think would be expected for a block support. Others might explain why this workaround was required 🤷 

Is it possible something was missed that results in the textAlign value being stripped before the attributes are saved? Or, can this behaviour be replicated with an existing block support and the test block?

Some other files that relate to block supports that might need updating:
- [packages/blocks/src/api/constants.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/constants.js?plain=1#L17) - this one could be important
- [hooks/support.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/supports.js)
- [lib/class-wp-theme-json-gutenberg.php](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php)
- [hooks/utils.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/utils.js)
- [global-styles/hooks.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/global-styles/hooks.js?plain=1#L24)
- [schemas/json/theme.json
](https://github.com/WordPress/gutenberg/blob/trunk/schemas/json/theme.json?plain=1#L554)
- [how-to-guides/themes/global-settings-and-styles.md](https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/themes/global-settings-and-styles.md?plain=1#L268)
- [reference-guides/theme-json-reference/theme-json-living.md](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/theme-json-reference/theme-json-living.md?plain=1#L271)

Others that might be out of scope for this support initially:
- [Global Styles typography panel](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/global-styles/typography-panel.js)
- Style engine e.g. [types](https://github.com/WordPress/gutenberg/blob/trunk/packages/style-engine/src/types.ts?plain=1#L57)

Hope that helps some 🤞